From 4f8a93f0653f537e45a752d196d9497d07dc397d Mon Sep 17 00:00:00 2001
From: Phil Edworthy <phil.edworthy@renesas.com>
Date: Mon, 10 Dec 2012 14:37:03 +0000
Subject: [PATCH 104/106] ARM: shmobile: r8a7779: Enable PMU (Performance
 Monitoring Unit)

This patch enables the PMU (Performance Monitoring Unit) for r8a7779.

Signed-off-by: Phil Edworthy <phil.edworthy@renesas.com>
---
 arch/arm/configs/marzen_defconfig      |    1 +
 arch/arm/mach-shmobile/setup-r8a7779.c |   32 ++++++++++++++++++++++++++++++++
 2 files changed, 33 insertions(+), 0 deletions(-)

diff --git a/arch/arm/configs/marzen_defconfig b/arch/arm/configs/marzen_defconfig
index f71abee..d2cc274 100644
--- a/arch/arm/configs/marzen_defconfig
+++ b/arch/arm/configs/marzen_defconfig
@@ -13,6 +13,7 @@ CONFIG_CGROUPS=y
 CONFIG_CC_OPTIMIZE_FOR_SIZE=y
 CONFIG_SYSCTL_SYSCALL=y
 CONFIG_EMBEDDED=y
+CONFIG_PERF_EVENTS=y
 CONFIG_SLAB=y
 CONFIG_MODULES=y
 CONFIG_MODULE_FORCE_LOAD=y
diff --git a/arch/arm/mach-shmobile/setup-r8a7779.c b/arch/arm/mach-shmobile/setup-r8a7779.c
index be89413..a9dc514 100644
--- a/arch/arm/mach-shmobile/setup-r8a7779.c
+++ b/arch/arm/mach-shmobile/setup-r8a7779.c
@@ -39,6 +39,7 @@
 #include <asm/mach/time.h>
 #include <asm/mach/map.h>
 #include <asm/hardware/cache-l2x0.h>
+#include <asm/pmu.h>
 
 static struct map_desc r8a7779_io_desc[] __initdata = {
 	/* 2M entity map for 0xf0000000 (MPCORE) */
@@ -311,6 +312,36 @@ static struct platform_device i2c3_device = {
 	.num_resources	= ARRAY_SIZE(rcar_i2c3_res),
 };
 
+static struct resource pmu_resources[] = {
+	[0] = {
+		.start	= gic_spi(59),
+		.end	= gic_spi(59),
+		.flags  = IORESOURCE_IRQ,
+	},
+	[1] = {
+		.start	= gic_spi(60),
+		.end	= gic_spi(60),
+		.flags  = IORESOURCE_IRQ,
+	},
+	[2] = {
+		.start	= gic_spi(61),
+		.end	= gic_spi(61),
+		.flags  = IORESOURCE_IRQ,
+	},
+	[3] = {
+		.start	= gic_spi(62),
+		.end	= gic_spi(62),
+		.flags  = IORESOURCE_IRQ,
+	},
+};
+
+static struct platform_device pmu_device = {
+	.name		= "arm-pmu",
+	.id		= ARM_PMU_DEVICE_CPU,
+	.num_resources = ARRAY_SIZE(pmu_resources),
+	.resource = pmu_resources,
+};
+
 /* SGX */
 static struct resource sgx_resources[] = {
 	[0] = {
@@ -350,6 +381,7 @@ static struct platform_device *r8a7779_early_devices[] __initdata = {
 };
 
 static struct platform_device *r8a7779_late_devices[] __initdata = {
+	&pmu_device,
 	&sgx_device,
 };
 
-- 
1.7.5.4

